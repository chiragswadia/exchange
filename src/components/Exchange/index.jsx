import React, { useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { NotificationContainer } from 'react-notifications'
import LiveRate from '../LiveRate'
import SourceCurrencySelector from '../SourceCurrencySelector'
import DestinationCurrency from '../DestinationCurrency'
import { calculateExchangeRate } from '../../helpers'
import * as actions from '../../actions'
import isValidTransaction from './utils/isValidTransaction'

import cn from './styles.module.css'

const RATE_UPDATE_INTERVAL = 600000 // TODO - Change this to 10000 before delivering

const Exchange = () => {
  const dispatch = useDispatch()
  const wallet = useSelector(state => state.wallet) // TODO - Maybe use memoized selectors for all if possible
  const rates = useSelector(state => state.rates)
  const amount = useSelector(state => state.form.amount)
  const sourceCurrency = useSelector(state => state.form.sourceCurrency)
  const destinationCurrency = useSelector(
    state => state.form.destinationCurrency
  )
  const exchangeRate = calculateExchangeRate({
    sourceCurrency,
    destinationCurrency,
    rates,
  })
  const isTransactionAllowed = isValidTransaction({
    sourceCurrency,
    destinationCurrency,
    wallet,
    amount,
    rates,
  })

  const performTransaction = event => {
    event.preventDefault()
    dispatch(actions.performTransaction())
  }

  const handleSourceCurrencyChange = sourceCurrency => {
    dispatch(actions.changeForm({ sourceCurrency }))
  }

  const handleDestinationCurrencyChange = destinationCurrency => {
    dispatch(actions.changeForm({ destinationCurrency }))
  }

  const handleAmountChange = amount => {
    dispatch(actions.changeForm({ amount }))
  }

  useEffect(() => {
    dispatch(actions.fetchAndUpdateRates())

    const rateUpdateIntervalId = setInterval(() => {
      dispatch(actions.fetchAndUpdateRates())
    }, RATE_UPDATE_INTERVAL)

    // Unmount
    return () => {
      clearInterval(rateUpdateIntervalId)
    }
  }, [dispatch])

  return (
    <form className={cn.root} onSubmit={performTransaction}>
      <NotificationContainer />
      <div className={cn.header}>
        <LiveRate
          sourceCurrency={sourceCurrency}
          destinationCurrency={destinationCurrency}
          rate={exchangeRate}
        />

        <button
          type="submit"
          disabled={!isTransactionAllowed}
          className={cn.button}
        >
          Exchange
        </button>
      </div>
      <div>
        <SourceCurrencySelector
          wallet={wallet}
          amount={amount}
          sourceCurrency={sourceCurrency}
          onChange={handleSourceCurrencyChange}
          onInput={handleAmountChange}
        />
        <DestinationCurrency
          sourceCurrency={sourceCurrency}
          destinationCurrency={destinationCurrency}
          wallet={wallet}
          amount={amount}
          rate={exchangeRate}
          onChange={handleDestinationCurrencyChange}
        />
      </div>
    </form>
  )
}

export default Exchange
