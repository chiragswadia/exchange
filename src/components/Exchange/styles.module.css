.root {
    position: relative;
    margin: 0 auto;
    max-width: 420px;
    border: 10px solid #eceff1;
    border-bottom-width: 60px;
    margin-top: 50px;
    border-radius: 20px;
  }
  
.header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 16px 8px 20px;
}

.button {
  background-color: rgb(235, 0, 141);
  border: none;
  width: 100px;
  height: 45px;
  padding: 0!important;
  color: #fff;
  cursor: pointer;
  font-size: 16px;
  border-radius: 2em;
}

.button:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}