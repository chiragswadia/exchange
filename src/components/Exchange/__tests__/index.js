import React from "react";
import { cleanup } from '@testing-library/react';
import renderer from "react-test-renderer";
import Exchange from "../";

beforeEach(cleanup);

jest.mock('react-redux', () => ({
    useDispatch: () => jest.fn(),
    useSelector: jest.fn(fn => fn({
        wallet: {
            EUR: 100,
            GBP: 200,
            USD: 200,
        },
        rates: {
            EUR: 1,
            USD: 1.11,
            GBP: 0.89,
        },
        form: {
            amount: 50,
            sourceCurrency: "EUR",
            destinationCurrency: "GBP",
        }
    })),
}));

describe("Exchange", () => {
    test("Snapshot should match", () => {
        expect(
            renderer.create(
                <Exchange />
            ).toJSON()
        ).toMatchSnapshot();
    });
});
