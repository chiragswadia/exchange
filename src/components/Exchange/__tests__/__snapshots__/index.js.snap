// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Snapshot should match 1`] = `
<form
  className="root"
  onSubmit={[Function]}
>
  <div
    className="notification-container notification-container-empty"
  >
    <span />
  </div>
  <div
    className="header"
  >
    <div
      className="root"
      data-testid="live-rate"
    >
      <span
        className="liveRateIndicator"
      />
      <span
        data-testid="formatted-currency"
      >
        €
         
        1
      </span>
       = 
      <span
        data-testid="formatted-currency"
      >
        £
         
        0.89
      </span>
    </div>
    <button
      className="button"
      disabled={true}
      type="submit"
    >
      Exchange
    </button>
  </div>
  <div>
    <div
      className="root"
    >
      <div
        className="root"
      >
        <div>
          <div
            className="carousel carousel-slider"
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <button
              aria-label="previous slide / item"
              className="control-arrow control-prev control-disabled"
              onClick={[Function]}
              type="button"
            />
            <div
              className="slider-wrapper axis-horizontal"
              style={Object {}}
            >
              <ul
                className="slider animated"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "MozTransform": "translate3d(-100%,0,0)",
                    "MozTransitionDuration": "350ms",
                    "MsTransform": "translate3d(-100%,0,0)",
                    "MsTransitionDuration": "350ms",
                    "OTransform": "translate3d(-100%,0,0)",
                    "OTransitionDuration": "350ms",
                    "WebkitTransform": "translate3d(-100%,0,0)",
                    "WebkitTransitionDuration": "350ms",
                    "msTransform": "translate3d(-100%,0,0)",
                    "msTransitionDuration": "350ms",
                    "transform": "translate3d(-100%,0,0)",
                    "transitionDuration": "350ms",
                  }
                }
              >
                <li
                  className="slide"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      USD
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        $
                         
                        200.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide selected"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      EUR
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        €
                         
                        100.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      GBP
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        £
                         
                        200.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      USD
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        $
                         
                        200.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide selected"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      EUR
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        €
                         
                        100.00
                      </span>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
            <button
              aria-label="next slide / item"
              className="control-arrow control-next control-disabled"
              onClick={[Function]}
              type="button"
            />
            <ul
              className="control-dots"
            >
              <li
                aria-label="slide item 1"
                className="dot selected"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
                value={0}
              />
              <li
                aria-label="slide item 2"
                className="dot"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
                value={1}
              />
              <li
                aria-label="slide item 3"
                className="dot"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
                value={2}
              />
            </ul>
          </div>
        </div>
      </div>
      <div
        className="amountInputContainer"
      >
        <input
          autoFocus={true}
          className="amountInput"
          data-testid="amount-input"
          onChange={[Function]}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="root"
    >
      <div
        className="root"
      >
        <div>
          <div
            className="carousel carousel-slider"
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <button
              aria-label="previous slide / item"
              className="control-arrow control-prev control-disabled"
              onClick={[Function]}
              type="button"
            />
            <div
              className="slider-wrapper axis-horizontal"
              style={Object {}}
            >
              <ul
                className="slider animated"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "MozTransform": "translate3d(-200%,0,0)",
                    "MozTransitionDuration": "350ms",
                    "MsTransform": "translate3d(-200%,0,0)",
                    "MsTransitionDuration": "350ms",
                    "OTransform": "translate3d(-200%,0,0)",
                    "OTransitionDuration": "350ms",
                    "WebkitTransform": "translate3d(-200%,0,0)",
                    "WebkitTransitionDuration": "350ms",
                    "msTransform": "translate3d(-200%,0,0)",
                    "msTransitionDuration": "350ms",
                    "transform": "translate3d(-200%,0,0)",
                    "transitionDuration": "350ms",
                  }
                }
              >
                <li
                  className="slide"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      USD
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        $
                         
                        200.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      EUR
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        €
                         
                        100.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide selected"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      GBP
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        £
                         
                        200.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      USD
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        $
                         
                        200.00
                      </span>
                    </div>
                  </div>
                </li>
                <li
                  className="slide"
                  onClick={[Function]}
                >
                  <div
                    className="root"
                  >
                    <div
                      className="currency"
                    >
                      EUR
                    </div>
                    <div
                      className="walletBalance"
                      data-testid="wallet-balance"
                    >
                      You have 
                      <span
                        data-testid="formatted-currency"
                      >
                        €
                         
                        100.00
                      </span>
                    </div>
                  </div>
                </li>
              </ul>
            </div>
            <button
              aria-label="next slide / item"
              className="control-arrow control-next control-disabled"
              onClick={[Function]}
              type="button"
            />
            <ul
              className="control-dots"
            >
              <li
                aria-label="slide item 1"
                className="dot"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
                value={0}
              />
              <li
                aria-label="slide item 2"
                className="dot selected"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
                value={1}
              />
              <li
                aria-label="slide item 3"
                className="dot"
                onClick={[Function]}
                onKeyDown={[Function]}
                role="button"
                tabIndex={0}
                value={2}
              />
            </ul>
          </div>
        </div>
      </div>
      <div
        className="amount"
      />
    </div>
  </div>
</form>
`;
